base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -880,85 +880,125 @@ where B: BlockchainBackend
     ///
     /// If an error does occur while writing the new block parts, all changes are reverted before returning.
     pub fn add_block(&self, candidate_block: Arc<Block>) -> Result<BlockAddResult, ChainStorageError> {
+        let timer = Instant::now();
+
+        let block_hash = candidate_block.hash();
         if self.is_add_block_disabled() {
             warn!(
                 target: LOG_TARGET,
                 "add_block is disabled, node busy syncing. Ignoring candidate block #{} ({})",
                 candidate_block.header.height,
-                candidate_block.hash(),
+                block_hash,
             );
             return Err(ChainStorageError::AddBlockOperationLocked);
         }
+        {
```

Comment:
Perhaps we should set `is_add_block_disabled` at this point?

I know the intention here was to reduce the amount of time that we spend with the write lock, but the danger here is that we switch to header sync or block sync in between each of these write locks.

I would actually suggest putting the write locks back across all operations and keep the other improvements in this PR

Reply:
If we want to do it, we should do it in another PR not this one. 
But this should not be required  because, header sync sets `is_add_block_disabled` and is the only one to set it, 
Block sync and header sync can disable it, But they will wait for the write lock here.

The only reason we have this here, and not just use the write lock, is because headersync and block sync has its own write lock each